### PR TITLE
flow trigger service

### PR DESCRIPTION
**Pull request by chengren311**
This PR added flow trigger service,  a singleton class in the AZ web server to process all flow trigger-related operations. Externally it provides following operations -
1. Create a trigger instance based on trigger definition.
2. Cancel a trigger instance.
3. Query running and historic trigger instances.
4. Recover incomplete trigger instances.
     
Internally, it
1. maintains the list of running trigger instance in memory.
2. updates status, starttime/endtime of trigger instance.
3. persists trigger instance to DB.
     
FlowTriggerService will be leveraged by Quartz scheduler, our new AZ scheduler to schedule starting triggers.
 
Flow Trigger Service internally contains a single-threaded executor service to process all trigger instance related operation to avoid complicated cases of race condition.

### CODE REVIEW

File: azkaban-web-server/src/test/java/azkaban/flowtrigger/util/TestUtil.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    final Map<String, String> props = new HashMap<>();
+    props.put(TestDependencyCheck.RUN_TIME, String.valueOf(runtimeInSec));
+    props.put(TestDependencyCheck.FAILURE_FLAG, String.valueOf(boundToFail));
+    return new FlowTriggerDependency(name, "TestDependencyCheck", props);
+  }
+
+  public static FlowTrigger createTestFlowTrigger(final List<FlowTriggerDependency> deps,
+      final Duration
+          maxWaitDuration) {
+    final FlowTrigger flowTrigger = new FlowTrigger(
```
Comment (kunkun-tang) [inline azkaban-web-server/src/test/java/azkaban/flowtrigger/util/TestUtil.java]: The two lines can be merged into one.

File: azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+    return this.runningTriggers.stream()
+        .flatMap(triggerInst -> triggerInst.getDepInstances().stream()).filter(
+            depInst -> depInst.getContext() != null && depInst.getContext() == context)
+        .findFirst().orElse(null);
+  }
+
+  /**
+   * Mark the dependency instance context as success
+   */
+  void markDependencySuccess(final DependencyInstanceContext context) {
```
Comment (kunkun-tang) [inline azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java]: mark this to public,

File: azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+
+  private boolean cancelledByDependencyPlugin(final DependencyInstance depInst) {
+    // When onKill is called by the dependency plugin not through flowTriggerService, we treat it
+    // as cancelled by dependency due to failure on dependency side. In this case, cancel cause
+    // remains unset.
+    return depInst.getStatus() == Status.CANCELLED && (depInst.getCancellationCause()
+        == CancellationCause.NONE);
+  }
+
+  void markDependencyCancelled(final DependencyInstanceContext context) {
```
Comment (kunkun-tang) [inline azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java]: make it to public please.

File: azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+    this.executorService.submit(() -> {
+      final DependencyInstance depInst = findDependencyInstanceByContext(context);
+      if (depInst != null) {
+        logger.info(
+            String.format("setting dependency instance[id: %s, name: %s] status to cancelled",
+                depInst.getTriggerInstance().getId(), depInst.getDepName()));
+        if (cancelledByDependencyPlugin(depInst)) {
```
Comment (kunkun-tang) [inline azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java]: don't understand this part

File: azkaban-web-server/src/test/java/azkaban/flowtrigger/util/TestUtil.java (hunk 1 above)
Comment (chengren311) [reply azkaban-web-server/src/test/java/azkaban/flowtrigger/util/TestUtil.java]: updated

File: azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java (hunk 3 above)
Comment (chengren311) [reply azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java]: updated

File: azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java (hunk 2 above)
Comment (chengren311) [reply azkaban-web-server/src/main/java/azkaban/flowtrigger/FlowTriggerService.java]: updated

**Pull request merged**